### PR TITLE
fix(dop,pipeline): Deleting pipeline definitions and sources when deleting application

### PR DESCRIPTION
**Pull request by littlejiancc**
#### What this PR does / why we need it:
Deleting pipeline definitions and sources when deleting project and application

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMjk3XSwiYXNzaWduZWUiOlsiMTAwMTI2MSJdfQ%3D%3D&id=315826&iterationID=1297&pId=0&type=BUG)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |     Deleting pipeline definitions and sources when deleting application         |
| 🇨🇳 中文    |      删除应用时删除流水线定义        |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: internal/tools/pipeline/providers/definition/db/define.go [hunk 1]
```diff
@@ -77,6 +78,16 @@ func (client *Client) DeletePipelineDefinition(id string, ops ...mysqlxorm.Sessi
 	return err
 }
 
+func (client *Client) DeletePipelineDefinitionByRemote(remote string, ops ...mysqlxorm.SessionOption) error {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	_, err := session.Table(new(PipelineDefinition)).
+		Where("remote LIKE ?", remote+"%").Where("soft_deleted_at = 0").
+		Update(map[string]interface{}{"soft_deleted_at": time.Now().UnixNano() / 1e6})
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/db/define.go:87]: use `time.Now().UnixMilli())`

File: .erda/migrations/pipeline/20220629-pipeline-definition-add-remote.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+ALTER TABLE `pipeline_definition` ADD COLUMN  `remote` varchar(100) NOT NULL DEFAULT '' COMMENT 'remote';
```
Comment (sfwn) [inline .erda/migrations/pipeline/20220629-pipeline-definition-add-remote.sql:1]: use `pipeline_source_id` is enough.

File: .erda/migrations/pipeline/20220629-pipeline-definition-add-remote.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+ALTER TABLE `pipeline_definition_extra` ADD COLUMN  `remote` varchar(100) NOT NULL DEFAULT '' COMMENT 'remote';
```
Comment (sfwn) [inline .erda/migrations/pipeline/20220629-pipeline-definition-add-remote.sql:2]: `pipeline_definition_extra` link `pipeline_source_id`.

**Review by sfwn — CHANGES_REQUESTED** [review]
stop cron when pipeline definition deleted

File: api/proto/core/pipeline/definition/definition.proto [hunk 4]
```diff
@@ -2,213 +2,231 @@ syntax = "proto3";
...
+    string sourceType = 14;
+    string remote = 15;
+    string ref = 16;
+    string path = 17;
+    string fileName = 18;
+    string pipelineSourceID = 19;
+    string status = 20;
+    int64 pipelineID = 21;
+    int64 totalActionNum = 22;
+    int64 ExecutedActionNum = 23;
```
Comment (sfwn) [inline api/proto/core/pipeline/definition/definition.proto:105]: why this field start with uppercase?

File: api/proto/core/pipeline/definition/definition.proto [hunk 5]
```diff
@@ -2,213 +2,231 @@ syntax = "proto3";
...
+    repeated string ref = 9;
+    repeated string remote = 10;
+    repeated string timeCreated = 11;
+    repeated string timeStarted = 12;
+    repeated string status = 13;
+    repeated string idList = 14;
+    repeated string ascCols = 15;
+    repeated string descCols = 16;
+    int64 totalActionNum = 17;
+    int64 ExecutedActionNum = 18;
```
Comment (sfwn) [inline api/proto/core/pipeline/definition/definition.proto:194]: and this

File: internal/apps/dop/endpoints/application.go [hunk 6]
```diff
@@ -168,6 +169,12 @@ func (e *Endpoints) DeleteApplication(ctx context.Context, r *http.Request, vars
 		return apierrors.ErrDeleteApplication.InternalError(err).ToResp(), nil
 	}
 
+	// delete project pipeline
+	_, err = e.ProjectPipelineSvc.DeleteByApp(ctx, &propipelinepb.DeleteByAppRequest{AppID: appDto.ID})
+	if err != nil {
+		logrus.Errorf("failed to delete projectPipeline by app, err: %v", err)
```
Comment (sfwn) [inline internal/apps/dop/endpoints/application.go:175]: abort app deletion, or project pipeline leaked.

File: internal/tools/pipeline/providers/definition/db/define.go [hunk 7]
```diff
@@ -77,6 +77,29 @@ func (client *Client) DeletePipelineDefinition(id string, ops ...mysqlxorm.Sessi
 	return err
 }
 
+func (client *Client) DeletePipelineDefinitionByRemote(remote string, ops ...mysqlxorm.SessionOption) error {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	_, err := session.Table("pipeline_definition").
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/db/define.go:84]: use model.TableName function

File: internal/tools/pipeline/providers/definition/db/define.go [hunk 8]
```diff
@@ -77,6 +77,29 @@ func (client *Client) DeletePipelineDefinition(id string, ops ...mysqlxorm.Sessi
...
+		Update(map[string]interface{}{"soft_deleted_at": time.Now().UnixNano() / 1e6})
+	return err
+}
+
+func (client *Client) ListPipelineDefinitionByRemote(remote string, ops ...mysqlxorm.SessionOption) ([]PipelineDefinitionSource, error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	var list []PipelineDefinitionSource
+	err := session.Table("pipeline_definition").Alias("d").
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/db/define.go:96]: same here

File: internal/tools/pipeline/providers/definition/db/define_extra.go [hunk 9]
```diff
@@ -57,6 +58,16 @@ func (client *Client) DeletePipelineDefinitionExtraExtra(id string, ops ...mysql
 	return err
 }
 
+func (client *Client) DeletePipelineDefinitionExtraByRemote(remote string, ops ...mysqlxorm.SessionOption) error {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	_, err := session.Table("pipeline_definition_extra").
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/db/define_extra.go:65]: same

File: internal/tools/pipeline/providers/definition/pipeline_definition.go [hunk 10]
```diff
@@ -170,6 +172,30 @@ func (p pipelineDefinition) Delete(ctx context.Context, request *pb.PipelineDefi
...
 }
 
+func (p pipelineDefinition) DeleteByRemote(ctx context.Context, request *pb.PipelineDefinitionDeleteByRemoteRequest) (*pb.PipelineDefinitionDeleteResponse, error) {
+	if request.Remote == "" {
+		return nil, fmt.Errorf("the remote is empty")
+	}
+
+	session := p.dbClient.NewSession()
+	defer session.Close()
+	session.Begin()
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/pipeline_definition.go:182]: handle error

File: internal/tools/pipeline/providers/definition/pipeline_definition.go [hunk 11]
```diff
@@ -170,6 +172,30 @@ func (p pipelineDefinition) Delete(ctx context.Context, request *pb.PipelineDefi
...
+
+	err := p.dbClient.DeletePipelineDefinitionByRemote(request.Remote, mysqlxorm.WithSession(session))
+	if err != nil {
+		session.Rollback()
```
Comment (sfwn) [inline internal/tools/pipeline/providers/definition/pipeline_definition.go:186]: rollback in defer

File: internal/tools/pipeline/providers/source/db/source.go [hunk 12]
```diff
@@ -130,6 +130,20 @@ func (client *Client) ListPipelineSource(idList []string, ops ...mysqlxorm.Sessi
 	return pipelineSource, nil
 }
 
+func (client *Client) ListPipelineSourceByRemote(remote string, ops ...mysqlxorm.SessionOption) ([]PipelineSource, error) {
+	session := client.NewSession(ops...)
+	defer session.Close()
+
+	var pipelineSource []PipelineSource
+	if err := session.Where("remote LIKE ?", remote+"%").
```
Comment (sfwn) [inline internal/tools/pipeline/providers/source/db/source.go:138]: let variable name clearer.
you'd better use `remotePrefix`.

File: internal/tools/pipeline/providers/source/pipeline_source.go [hunk 13]
```diff
@@ -96,6 +97,32 @@ func (p pipelineSource) Delete(ctx context.Context, request *pb.PipelineSourceDe
...
+		return nil, fmt.Errorf("the remote is empty")
+	}
+
+	sources, err := p.dbClient.ListPipelineSourceByRemote(request.Remote)
+	if err != nil {
+		return nil, err
+	}
+	session := p.dbClient.NewSession()
+	defer session.Close()
+	session.Begin()
```
Comment (sfwn) [inline internal/tools/pipeline/providers/source/pipeline_source.go:111]: same

**Review by sfwn — CHANGES_REQUESTED** [review]

File: api/proto/core/pipeline/definition/definition.proto (hunk 4 above)
Comment (littlejiancc) [reply api/proto/core/pipeline/definition/definition.proto:105]: done

File: internal/tools/pipeline/providers/definition/db/define.go (hunk 7 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/definition/db/define.go:84]: done

File: internal/tools/pipeline/providers/definition/db/define.go (hunk 8 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/definition/db/define.go:96]: done

File: internal/tools/pipeline/providers/definition/pipeline_definition.go (hunk 10 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/definition/pipeline_definition.go:182]: done

File: internal/tools/pipeline/providers/definition/db/define_extra.go (hunk 9 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/definition/db/define_extra.go:65]: done

File: internal/tools/pipeline/providers/definition/pipeline_definition.go (hunk 11 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/definition/pipeline_definition.go:186]: done

File: internal/tools/pipeline/providers/source/pipeline_source.go (hunk 13 above)
Comment (littlejiancc) [reply internal/tools/pipeline/providers/source/pipeline_source.go:111]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**